acquia/mc-cs-plugin-custom-objects
# Fix segment build when multiple custom items are linked to a contact

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [✅]
| New feature/enhancement? (use the a.x branch)      | [❌]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:


This PR fixes the way CO filters are used in segments. When multiple custom items are linked to a contact, it didn't use to account a single custom object row while considering segment filter query. This fix can be enabled by setting `custom_object_merge_filter` key to true and also `custom_object_item_value_to_contact_relation_limit` key to 0.

If the following contacts have custom items associated and consider today's date is *2024-02-12*:
| Contact ID | Custom Item Text | Custom Item Date |
| --- | --- | --- |
| 1 | Tiger | 2024-02-01 |
| 1 | Lion | 2024-02-01 |
| **2** | **Tiger** | **2024-02-15** |
| 2 | Lion | 2024-02-15 |
| 3 | Tiger | 2024-03-15 |
| 3 | Lion | 2024-03-15 |

Segment filters are:
1. Custom Item Text = 'Tiger'
2. Custom Item Date > '2024-02-13 23:59:59'
3. Custom Item Date <= '2024-03-12'

### Before PR:
```sql
SELECT l.id
FROM mautic_leads l
WHERE (EXISTS(SELECT contact_id
              FROM mautic_custom_field_value_text cfwq_2_value
                       INNER JOIN mautic_custom_item_xref_contact cfwq_2_contact
                                  ON cfwq_2_value.custom_item_id = cfwq_2_contact.custom_item_id
              WHERE (cfwq_2_value.custom_field_id = 2)
                AND (cfwq_2_value.value = 'Tiger')
                AND (cfwq_2_contact.contact_id = l.id)))
  AND (EXISTS(SELECT contact_id
              FROM mautic_custom_field_value_date cfwq_1_value
                       INNER JOIN mautic_custom_item_xref_contact cfwq_1_contact
                                  ON cfwq_1_value.custom_item_id = cfwq_1_contact.custom_item_id
              WHERE (cfwq_1_value.custom_field_id = 1)
                AND (cfwq_1_value.value > '2024-02-13 23:59:59')
                AND (cfwq_1_value.value <= '2024-03-12')
                AND (cfwq_1_contact.contact_id = l.id)))
  AND (l.id NOT IN (SELECT par5.lead_id FROM mautic_lead_lists_leads par5 WHERE par5.leadlist_id = 1))

```
The segment should only consider contact ID 2 has a valid contact but since the SQL generated individually executed consider contact IDs 1 and 3 as valid contacts.

### After PR:
```sql
SELECT l.id
FROM mautic_leads l
WHERE (EXISTS(SELECT 1
              FROM mautic_custom_item_xref_contact cix
                       INNER JOIN mautic_custom_field_value_text cix_2_text_value
                                  ON cix_2_text_value.custom_item_id = cix.custom_item_id AND
                                     cix_2_text_value.custom_field_id = 2
                       INNER JOIN mautic_custom_field_value_date cix_1_date_value
                                  ON cix_1_date_value.custom_item_id = cix.custom_item_id AND
                                     cix_1_date_value.custom_field_id = 1
              WHERE (cix.contact_id = l.id)
                AND (cix_2_text_value.value = 'Tiger')
                AND (cix_1_date_value.value > '2024-02-13 23:59:59')
                AND (cix_1_date_value.value <= '2024-03-12')))
  AND (l.id NOT IN (SELECT par4.lead_id FROM mautic_lead_lists_leads par4 WHERE par4.leadlist_id = 4))
```

#### Steps to test this PR:


1. Open this PR for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Try above scenario and it should consider only contact 2
